### ISSUE TITLE
GeoPandas performance: optimizing vectorized operations

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#430]
**STATUS UPDATE**: work on vectorized operations has been done in the PyGEOS package (to be integrated into Shapely), and optional support for this has landed in geopandas master in the meantime.  See https://github.com/geopandas/geopandas/issues/1155, and https://geopandas.readthedocs.io/en/latest/install.html#using-the-optional-pygeos-dependency on how to test this.

---

There have been some issues recently about performance of geopandas (eg https://github.com/geopandas/geopandas/issues/396), and once you have somewhat larger datasets this can be a problem. On the one hand, there are certainly algorithmic optimizations possible (more performant algorithm, better use of spatial indices, etc. For example the current work on `overlay` https://github.com/geopandas/geopandas/pull/338, https://github.com/geopandas/geopandas/pull/429). But, even then the basic operations can still be slow, making it difficult to ever reach really performant code.

---

The main problem with geopandas' performance is that all vectorized operations in geopandas are just wrappers around for-loops / list comprehensions. For example: 

```
geoseries.distance(point) 
```

just calls the ``distance`` method of the underlying shapely objects many times in a python loop.

So the question then is, can't we do this for loop on a lower level. Shapely is a wrapper around the GEOS C library using ctypes (and partly cython), so it could be possible to do the loops on this lower level. Or with some of the other tools (cython, cffi, numba, ..) to push down the for loop to a lower level, the same way as the performant operations in pandas happen.

So I thought to just try it out to explore what would be possible, and I tried it with cython (based on what is available in shapely of cython wrappers). Result of this experimentation can be seen here: http://nbviewer.jupyter.org/gist/jorisvandenbossche/2e38c8ae14d273d9d7b117318e23c7ed 
Summary, I get **a speed-up of ca 25x (and 250x when using integer `_geom` pointers)** on this specific example.

It certainly shows that large speed-ups on basic operations are possible. But, I am not an expert in both cython or geos, so I am not sure this is a good approach. But the way I did it here for 'contains' can certainly be expanded to the other methods, and it seems something worth to explore further.

**Some questions**

- Is this is a good approach? Is the way I use Cython here a good approach? (apart from that it can be cleaned-up by a more knowledgeable cython coder). Would something else be more appropriate / easier (numba, ctypes, cffi, ...)

- Is it generally safe to use those integers pointers? (`_geom` attribute)

- Where would this functionality belong? (cc @sgillies) 
  - There is already a `shapely.vectorized` module, and I think such vectorized operations for other methods would suite there as well (with the only difference that the current methods in `shapely.vectorized` do not work with arrays of geomtries, but with x and y arrays).
  - Including it in shapely would make it easier to distribute (it already has all machinery to handle the deps and compilation etc), but would possibly expand scope. Including it in geopandas would make it easier to experiment/expand things as needed, but would make the installation/maintenance harder, and duplicating code from shapely.
  
- Are there people who would like to collaborate on this?

cc @ozak @kuanb @gboeing

**Comment by ozak** [comment]
I'd be happy to help... Although I'm also a novice in most ``cython`` related matters. Unless `` numba`` had changed so that you can have compiled functions, ``jit`` may not be the way to go due to overhead. I was just recently looking at [this blog](https://www.ibm.com/developerworks/community/blogs/jfp/entry/Python_Meets_Julia_Micro_Performance?lang=en) that introduced some ways to improve python code's performance. It has done additional links that may be worth exploring.

**Comment by kuanb** [comment]
@jorisvandenbossche as mentioned over email I would be interested in helping but, like @ozak (great blog post btw, reading it rn thanks), must warn that I am similarly a novice with regard to `cython`.

**Comment by gboeing** [comment]
This is interesting. There'd be a lot of benefit in vectorizing or compiling those geopandas loops. For instance: I had to build a lot of hack-y spatial queries into OSMnx to speed up my spatial operations for working with millions of points and complex polygons.

It would be interesting to benchmark a numba solution against your cython solution, as it might be simpler to implement. You'd presumably have to fall back on numpy x-y arrays instead of shapely geometries to get the functions to compile with numba (since it only plays nicely with scalars and arrays). But new versions of numba might be able to compile classes.

For what it's worth, the existing shapely vectorized module alone provides an order of magnitude speed increase over the python-loops implementation in your example code:

```python
import numpy as np, shapely.vectorized as sv
from shapely.geometry import Point, Polygon
polygon = Polygon([(10,10), (10,100), (100,100), (100, 10)])
xy = np.array([(n, n) for n in range(10000)])
points = [Point(x, y) for x, y in xy]
```
Loops:
```python
def contains_py(polygon, points):
    return np.array([polygon.contains(point) for point in points])
%timeit contains_py(polygon, points)
```
```
10 loops, best of 3: 40.1 ms per loop
```

Vectorized:
```python
%timeit sv.contains(polygon, x=xy[:,0], y=xy[:,1])
```
```
100 loops, best of 3: 3.09 ms per loop
```

But integrating this with a spatial index probably opens a whole new can of worms.

**Comment by sgillies** [comment]
I'm 👍 on enhancements to Shapely that we can generalize to other applications.

Something to keep in mind: requiring multiple extension modules that use the same C shared library causes a dramatic increase in the complexity of distributing a project. If GeoPandas were to add an extension module that used GEOS, users will always have to make sure that the GeoPandas and Shapely extension modules were built in a compatible way (same Python SDKs, same macosx development target, etc).

**Comment by nickeubank** [comment]
I can't speak to some of @jorisvandenbossche questions about safety of using pointers, but I in general the use of Cython seems great. It's already such a backbone of Python datastack I can't imagine any issues. Also, I don't know how easily other tools (like numba) will integrate. I'd be happy to help collaborate if I can, though while I have some familiarity with C and I'm really a cython-and-up kinda programmer, so probably can't help much with the super low-level stuff. 

Re: using vectorized shapely functions: this seems a great first-pass, and maybe something I'll try and put together sooner rather than later -- seems like it'd be a relatively easy patch and frankly I could use the speedup in my own work anyway! :)

**Comment by nickeubank** [comment]
Interesting -- shapely.vectorized only seems to want to check points, which complicates Polygon / Polygon spatial joins...

One other point: looking at `sjoin`, it doesn't actually seem like we're taking full advantage of the index. Unless I'm missing something, the current strategy is (a) identify any observations in `left_df` that join with something, (b) identify any observations in `right_df` that join with something, then run `contains` or `intersects` on all pairwise combinations in those two residual dataframes.

I think something better would be something like:

        def get_joins(row):
            joins = list()
            for possible_match in idxmatch[row.name]:
                if row.prepared.intersects(right_df.iloc[possible_match].geometry):
                    joins.append(possible_match)
            return joins
        
        left_df['prepared'] = left_df.geometry.apply(prepared.prep)
        a = left_df[['prepared']].apply(get_merges, axis='columns')
 
that only uses shapely `contains` or `intersects` for objects identified as potential pair-wise matches by the rtree index.

**Comment by nickeubank** [comment]
Wow my little tweak is much slower. Never mind.

**Comment by jorisvandenbossche** [comment]
Thanks for the feedback all!

> I'm :+1: on enhancements to Shapely that we can generalize to other applications.

@sgillies That's great to hear. 
I had some presumption that also adding c extension in pandas against GEOS would complicate stuff, so that would indeed be good to keep all geos c interfaces gathered in shapely (and keeping geopandas pure python). 
I will open an issue at shapely to discuss more in detail.

> It would be interesting to benchmark a numba solution against your cython solution, as it might be simpler to implement. You'd presumably have to fall back on numpy x-y arrays instead of shapely geometries to get the functions to compile with numba (since it only plays nicely with scalars and arrays). But new versions of numba might be able to compile classes.

@gboeing I did an attempt to integrate with cffi and numba (but didn't post that here). The main problem here is that I (in my current attempt) wanted to keep using GEOS, to not have to reimplement the different geometrical predicates and functions in numba (that is what I am doing in the cython approach: just calling the same GEOS C functions as shapely does, only in a cython loop to overcome the python overhead). 
Numba has support for working with cffi, but I didn't yet figure out how to register a cffi-wrapped geometry type with numba, or whether this is even possible (just calling the GEOS function with CFFI worked, but not yet putting it in a numba function).

You could of course do the actual algorithm in numba as well, but that is much more work (or I am not familiar enough with that) 
To integrate with shapely, cython seems the better fit, but it would still be interesting to further look into numba for user defined functions.

**Comment by ozak** [comment]
@jorisvandenbossche I looked a bit at ``numba`` and it seems one would have to create all the classes in ``numba`` first. So, this may not be a good avenue, unless there is a way to import all the classes from ``shapely/geopandas`` into ``numba`` somehow.

**Comment by ozak** [comment]
Found [this jupyter notebook](http://nbviewer.jupyter.org/github/synapticarbors/rmath-cffi-example/blob/master/rmath-cffi-example.ipynb), which shows how to use C libraries from within ``numba``...I think @jorisvandenbossche mentioned looking into something like this (could you share some of the code?). This is a quite more advanced than what I know at the moment, but perhaps others can follow up....or if I can see some of the code you played with I may be able to help a bit more.

**Comment by wmay** [comment]
Whatever happened to this? I'd also be happy to help (and I'm also new to cython).

**Comment by jorisvandenbossche** [comment]
@ozak in the meantime, I actually managed to get a speed-up for vectorized methods with numba. I didn't get it working using CFFI, but after that I tried with the ctypes wrapper (which is already available from shapely), and that was actually rather easy: http://nbviewer.jupyter.org/gist/jorisvandenbossche/0235c96f4c41c8b462a2ee48ddfb268d (notebook is not yet cleaned up, sorry) 
End result is more or less the same speed-up as using cython (only with a bit less boilerplate, this is in general my experience if you optimize both ways you get more or less the same speedup).

But for integrating with shapely, and more easily distributing it etc, I still think cython is the more appropriate option for now (using numba in geopandas would require users to also have numba installed, which is something we would like to avoid I think. Although it could maybe be used for certain speed-ups that are optional).

**Comment by jorisvandenbossche** [comment]
> So, to implement the cython solution you showed would require rewriting everything in base.py? Or is there a way to define cython version of _series_op, _geo_op, _geo_unary_op, and _series_unary_op that would take of this?

@ozak @wmay the first step would actually be to contribute the vectorized methods (like the example in the cython notebook) to shapely. 

And then in a second step we can start using that (conditionally on the shapely version) in geopandas. So in the operations in `base.py`, where we now typically have list comprehensions like `[getattr(s, op)(other) for s in this.geometry]`, we should make another version using something like `shapely.vectorized.<op>(s, other)`, and just execute one of the two depending on the shapely version.

**Comment by mrocklin** [comment]
Pinging (nerd sniping) @eriknw

**Comment by mrocklin** [comment]
Are all relevant vectorizable operations within shapely or are there other operations, like plotting or joins that might require cython at the geopandas side?

**Comment by eriknw** [comment]
I bet I could help with Cythonizing packages, but I don't know much about `geopandas` or `shapely`, so I doubt I could drive this forward on my own.

Is anybody here going to the SciPy conference next week?  I am, so maybe we could discuss this in person or partner up and work on it.

**Comment by wmay** [comment]
I opened an issue for shapely with a possible roadmap for making this happen, https://github.com/Toblerity/Shapely/issues/501.

**Comment by mrocklin** [comment]
The intent of my question here:

> Are all relevant vectorizable operations within shapely or are there other operations, like plotting or joins that might require cython at the geopandas side?

Was to see if this work should happen in shapely or in geopandas.  If not everything can be done in shapely then it might make sense to focus vectorization work in geopandas.

**Comment by jorisvandenbossche** [comment]
@mrocklin there is the concern raised in https://github.com/geopandas/geopandas/issues/430#issuecomment-291083993 that it will complicate building / installation a lot if geopandas also depends on the C geos. 
Did your recent experiment highlight some functions that could use vectorization but would not suite in shapely?

**Comment by mrocklin** [comment]
From my perspective the recent VectorizedGeometry work could go into either library.  I'm doing my work in this branch just because I started here, but I would be happy to move it over to shapely if everyone agrees that that's best.  

However, GeoPandas might want to take on Cython and linking to GEOS regardless for operations like spatial joins, which may be out of scope for shapely (although I'm not sure where to draw the line here).

**Comment by mrocklin** [comment]
There are a few questions that I think need the attention of GeoPandas and Shapely developers:

1.  Do we want to manage GEOSGeometry objects outside of shapely objects?
2.  If (1) then where should this live?  GeoPandas or Shapely?
3.  If (1) then how do we handle the `'geometry'` column in a GeoDataFrame?  How do we convince Pandas to allow this new object within a dataframe?  We could just move around the pointers without a wrapper class.  We could attach the geometries as an attribute rather as a normal pandas column.  Perhaps we could subclass ndarray in some way to avoid coercion.

cc @jorisvandenbossche @sgillies 

I have such an implementation here: https://github.com/mrocklin/geopandas/blob/cythonize/geopandas/_vectorized.pyx in https://github.com/geopandas/geopandas/pull/467

**Comment by jorisvandenbossche** [comment]
All, an initial implementation has been merged (see #467), including vectorized geometry operations in cython based on C pointers, and a GeometryBlock to incorporate this into a GeoSeries and GeoDataFrame. 
This work should give a big performance boost, make it possible to add some more improvements in algos (@mrocklin already did some work on `sjoin`) and io functions, and make it easier to parallelize / integrate with dask.

Cool stuff, but still rough, and now it could use feedback from actual users! So you are all very welcome to check it out, test it on some of your application, and give feedback.

I will provide some more detailed instructions on how to check it out in https://github.com/geopandas/geopandas/issues/473

**Comment by jorisvandenbossche** [comment]
A bit more explanation of the current architecture: we made a vectorized geometry object `GeometryArray` (array-like with vectorized operations) in cython in geopandas. This vectorized geometry object only holds the integer pointers as its data, and only boxes it to shapely objects when the user accesses eg a single element, or iterates over it, ... This makes it fast and cheaper to construct. 

To integrate this in the GeoDataFrame and GeoSeries, we implemented a new GeometryBlock ('blocks' are the internal building block of pandas for the different columns). The reason we need a custom GeometryBlock, is because we need to have a way to let pandas know the data are not just normal integers we store in the dataframe (it are pointers to geometry objects), and cannot be manipulated as it were integers.

I updated #473 with some basic instructions to test this.

**Comment by jorisvandenbossche** [comment]
TLDR: we are working on the `pygeos` package that provides vectorized GEOS operations and will provide a performance boost to GeoPandas.

---

An update on this (after a while ..): although I updated the cython branch earlier this summer (https://github.com/geopandas/geopandas/pull/1030) to bring it up to date with the refactored internals in version 0.6.0 (based on ExtensionArrays, this refactor already brought parts of the cython branch into master), effort has now switched to the `pygeos` package (https://github.com/caspervdw/pygeos/), which will provide the same speed-ups but is using a different (and more general) approach. 
It was a happy coincidence that @caspervdw started working on this package at more or less the same time that I started looking again at this performance issue in geopandas!

Short summary of **`pygeos`**: it provides vectorized GEOS operations on numpy arrays of geometries, mostly using the numpy ufunc machinery. This means that we can use it in GeoPandas for performing the operations on the 1D geometry column, but it can also work in general on numpy arrays and make use of numpy broadcasting etc (eg to automatically calculate all intersections or distances between combinations of 2 arrays giving a 2D array as result). 
It implements a lightweight geometry python extension type wrapping a GEOSGeometry pointer (which knows how to deallocate itself). This makes it more robust as storing the raw pointers as in the cython appraoch, while the python extension type's struct with the pointer is still accessible from C/Cython without much overhead.

So, we can use pygeos in GeoPandas to achieve a performance boost on the geospatial operations.

Some of the questions that have been discussed above are still relevant, such as: does this belong in Shapely itself, or outside (now in pygeos instead of geopandas)? If not, how do we ensure compatibility (both on a packaging level, as for user interfacing geometry objects) ? 
That are still aspects that need be further discussed.

One specific aspect for geopandas to discuss is how to integrate with pygeos initially (require it, do it optional for now, ...), but will open a specific issue for that.

**Comment by mrocklin** [comment]
That's an exciting development.  I'm glad to see progress here.  Thank you for the solution @caspervdw !

**Comment by martinfleis** [comment]
With pygeos support in geopandas and shapely 2.0 on the way, I think that it is time to close this issue. Thanks all!
